RavenDB-20628 - Handle failover on cluster transaction (6.0)

shaharhikri opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20628/Cluster-wide-transaction-throw-TaskCanceledException-when-fails-on-timeout

### Additional description

Cluster wide transaction : throw `TaskCanceledException` when fails on timeout (happens in failover) - 6.0
The scenario:
The client sends the request to node A, it finishes performing the cluster transaction, then it fails.
Now client sends the same request to node B (as part of the failover) but it throws `TaskCanceledException`.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

shaharhikri committed:
RavenDB-20628 - Cluster wide transaction : throw TaskCanceledException when fails on timeout (6.0)

shaharhikri committed:
RavenDB-20628 - Fix FromRemote for DynamicJsonArray

shaharhikri committed:
RavenDB-20628 - Move the deserialization from JsonDeserializationCluster to ClusterTransactionCommand (for FromRemote method)

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -753,21 +765,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
+                        Only then we will want the task that created here, because we'll want to 'wait for index' in the new node after the failover.
+                    */
+
+                    indexTask.ContinueWith(t =>
```
Avoid throwing here, you can check via the `IsCompletedSucessfully`

shaharhikri replied:
Done, I preferred to check if it's faulted

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -753,21 +765,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
-                }
 
-                var result = new ClusterTransactionCompletionResult
+                    var removeTask = ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(command.Options.TaskId);
```
Why `removeTask`? What is being removed here?

shaharhikri replied:
this task is held by the ClusterTransactionWaiter and its completion tells us that the cluster transaction is completed (finished in the database).
The 'remove task' is IDisposable that at his disposal removes a cluster transaction task from the ClusterTransactionWaiter.
At the end of the cluster transaction on the database, if we need to 'wait for indexing' we create a task in the ClusterTransactionWaiter and set the task to be completed only when the indexing (task) is completed.
The task won't be created if it's already in the ClusterTransactionWaiter (created somewhere else),
in that case we will get the existing task when we'll call the ClusterTransactionWaiter.CreateTask.
We will remove the task from the ClusterTransactionWaiter dictionary, in assumption that in the places that we should 'await' that task, we already got its instance before it was removed from the ClusterTransactionWaiter dict (by disposing of its 'remove task').
We remove it here because it can cause a memory leak..
Creating the task here is adding it to the ClusterTransactionWaiter dict, and we cannot be sure that it will be removed from the dict somewhere else.

In case we don't 'wait for index' we will set the task only if it has being created in somewhere else in the ClusterTransactionWaiter (it is usually created in the places where we need to wait for it).

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -753,21 +765,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
+
+                    indexTask.ContinueWith(t =>
+                    {
+                        try
+                        {
+                            t.GetAwaiter().GetResult(); // Task t is already completed here
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, null);
+                            ServerStore.Cluster.ClusterTransactionWaiter.SetResult(options.TaskId, taskResult);
+                        }
+                        catch (Exception e)
+                        {
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, e);
+                            ServerStore.Cluster.ClusterTransactionWaiter.SetException(options.TaskId, e);
+                        }
```
Weren't we already calling that ealier?

shaharhikri replied:
no, we set the task only in the end of the indexing or (xor) in its completion in the database (if we don't need to wait form indexing).
It happens only once.

shaharhikri replied:
It happens when the cluster transaction raft command is being committed only when the database is shared (that was the behavior before my changes).

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -790,7 +830,7 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
 
                 RachisLogIndexNotifications.NotifyListenersAbout(index, exception);
-                ServerStore.Cluster.ClusterTransactionWaiter.TrySetException(options.TaskId, exception);
+                ServerStore.Cluster.ClusterTransactionWaiter.SetException(options.TaskId, exception);
```
This will throw, no?

shaharhikri replied:
No, because we will set it only once.

shaharhikri replied:
It happens when the cluster transaction raft command is being committed only when the database is shared (that was the behavior before my changes).

ayende replied:
Pretty sure that you cannot guarantee that.
You may have a race condition with waiting for indexing already calling this.

shaharhikri replied:
Done

shaharhikri committed:
RavenDB-20628 - Remove trying to get cluster transaction result from ClusterTransactionWaiter task and log history

shaharhikri committed:
RavenDB-20628 - Fix failing tests

shaharhikri committed:
RavenDB-20628 - move out errors list from ClusterTransactionResult for old follower (and updated leader)

shaharhikri committed:
RavenDB-20628 - Add interversion tests

shaharhikri committed:
RavenDB-20628 - Fix failing tests

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -786,8 +786,9 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
                         {
-                            RachisLogIndexNotifications.NotifyListenersAbout(index, t.Exception);
-                            ServerStore.Cluster.ClusterTransactionWaiter.SetException(options.TaskId, t.Exception);
+                            Exception e = t.Exception.InnerException;
```
Why are you assuming that there is an inner exception here? What if there isn't?

shaharhikri replied:
Done

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -457,6 +458,13 @@ record = _serverStore.Cluster.ReadDatabase(context, Name, out index);
                     try
                     {
+                        using (DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext ctx))
+                        using (ctx.OpenReadTransaction())
+                        {
+                            var lastCompletedClusterTransactionIndex = DocumentsStorage.ReadLastCompletedClusterTransactionIndex(ctx.Transaction.InnerTransaction);
+                            Interlocked.Exchange(ref LastCompletedClusterTransactionIndex, lastCompletedClusterTransactionIndex);
+                        }
```
Why are you doing that in the background thread?
This is a one time operation that should go in the database init. 
Then you don't need to worry about concurrency

shaharhikri replied:
Done

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -753,21 +764,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
-                }
 
-                var result = new ClusterTransactionCompletionResult
+                    var removeTask = ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(command.Options.TaskId);
```
I don't think that this is needed at all.
Why do you _care_ about this separately? You don't care about the task result here any longer, do you?
That is handled by the batch handler

shaharhikri replied:
`ClusterTransactionWaiter.CreateTask` will create a task if it doesn't exist, and if it exists, it will take it instead of creating a new one. 

If we remove the `ClusterTransactionWaiter.CreateTask` from here, it can create some race in case of 'wait for index'.

between this line:
https://github.com/ravendb/ravendb/blob/1a4a13a3f6176e1a6e08be66fe0dfc1519bad8e9/src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs#L68
and this line:
https://github.com/ravendb/ravendb/blob/1a4a13a3f6176e1a6e08be66fe0dfc1519bad8e9/src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs#L71

Suppose between those lines the cluster transaction is already completed in the database (the `LastCompletedClusterTransactionIndex` is equal to or greater than the cluster transaction index) but the indexing wasn't completed yet.
In that case, it won't wait for the indexing.

ayende replied:
Then move the `CreateClusterTransactionTask` _before_ the `SendToLeaderAsync`, no?

shaharhikri replied:
You cant, because you need the pass the raft command index to create the task

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -753,21 +764,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
+                        }
+                        else
+                        {
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, null);
```
This is wrong, assume that there is _another_ notification about a *later* index while we are waiting for the indexes to complete, what happens then?

shaharhikri replied:
Done, I moved it out from the 'if else' scopes

shaharhikri committed:
RavenDB-20628 - Addressed PR Comments

shaharhikri committed:
RavenDB-20628 - Move SendToLeaderAsync after creating the cluster tx task

ayende commented on src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs:
```diff
@@ -43,72 +46,88 @@ public async ValueTask<(long Index, DynamicJsonArray Results)> ProcessAsync(Json
-            }
+    private async Task<DynamicJsonArray> GetClusterTransactionDatabaseCommandsResults(object result, long databaseCommandsCount, long index, Task<Task> onDatabaseCompletionTask)
+    {
+        // old leader returned errors
```
note

shaharhikri replied:
Deleted, This comment isn't relevant anymore..

ayende commented on src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs:
```diff
@@ -43,72 +46,88 @@ public async ValueTask<(long Index, DynamicJsonArray Results)> ProcessAsync(Json
+            return clusterTxResult.GeneratedResult;
         }
+
+        // old leader returned null
```
note

shaharhikri replied:
Deleted

ayende commented on src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs:
```diff
@@ -43,72 +46,88 @@ public async ValueTask<(long Index, DynamicJsonArray Results)> ProcessAsync(Json
-            var raftRequestId = RequestHandler.GetRaftRequestIdFromQuery();
-            ClusterTransactionCommand clusterTransactionCommand = CreateClusterTransactionCommand(parsedCommands, options, raftRequestId);
+        ClusterTransactionCommand clusterTransactionCommand = CreateClusterTransactionCommand(parsedCommands, options, raftRequestId);
+        DynamicJsonArray array;
+        object result;
+        long index;
```
reoder

shaharhikri replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/Processors/Batches/ClusterTransactionRequestProcessor.cs:
```diff
@@ -22,6 +27,18 @@ public ClusterTransactionRequestProcessor(DatabaseRequestHandler requestHandler,
+            var lastCompleted = Interlocked.Read(ref RequestHandler.Database.LastCompletedClusterTransactionIndex);
+            if (lastCompleted >= index)
+            {
+                return Task.FromResult(Task.CompletedTask);
```
here we aren't waiting for the index task

ayende replied:
Here we need to check if we have to wait for index, and then wait for the indexes to complete as of the current etag.
Need to only account for indexes modified by the actions in the tx

ayende replied:
Better yet, move the index wait entirely to here instead of the tx command

shaharhikri replied:
Done

ayende commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Batches/ShardedClusterTransactionRequestProcessor.cs:
```diff
@@ -15,9 +19,20 @@ public ShardedClusterTransactionRequestProcessor(ShardedDatabaseRequestHandler r
+    public override Task<Task> WaitForDatabaseCompletion(Task<Task> onDatabaseCompletionTask, long index, CancellationToken token)
+    {
+        token.ThrowIfCancellationRequested();
+
+        if (onDatabaseCompletionTask.IsCompletedSuccessfully)
+            return onDatabaseCompletionTask;
+
+        // failover
```
Explain this properly in a comment

shaharhikri replied:
Deleted

ayende commented on src/Raven.Server/Rachis/ClusterTransactionWaiter.cs:
```diff
@@ -2,24 +2,23 @@
-        public DynamicJsonArray Array;
+        public RemoveTask CreateTask(string id, out Task<Task> task)
+        {
+            task = _results.GetOrAdd(id, new TaskCompletionSource<Task>(TaskCreationOptions.RunContinuationsAsynchronously)).Task;
```
```suggestion
            task = _results.GetOrAdd(id, static () =>  new TaskCompletionSource<Task>(TaskCreationOptions.RunContinuationsAsynchronously)).Task;
```

shaharhikri replied:
Done

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -661,14 +682,28 @@ private struct SingleCommandForShard
                         return;
 
-                    clusterTransactionWaiter.TrySetResult(Options.TaskId, new ClusterTransactionCompletionResult { Array = result });
+                    clusterTransactionWaiter.TrySetResult(Options.TaskId, Task.CompletedTask);
```
Remove this

shaharhikri replied:
Done

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -891,20 +926,76 @@ private static unsafe SingleClusterDatabaseCommand ReadCommand<TTransaction>(Tra
+            return array;
+        }
+
+        private static DynamicJsonValue ToDynamicJsonValue(BlittableJsonReaderObject bjro)
```
See if we can do:

```

    public class Result
    {
        public string Type;
        [ForceJsonSerialization("@change-Vector")]
        public string ChangeVector;
    }
    ```

shaharhikri replied:
Need to open another issue for that after this PR is merged

karmeli87 replied:
please opene an issue for that

ayende replied:
what is the issue number?

shaharhikri replied:
[RavenDB-21984](https://issues.hibernatingrhinos.com/issue/RavenDB-21984)

shaharhikri committed:
RavenDB-20628 - Addressed PR Comments

shaharhikri committed:
RavenDB-20628 - Register cluster tx and indexing tasks into the right token

shaharhikri commented:
test this please

shaharhikri commented:
test this please

karmeli87 commented on src/Raven.Server/Documents/Handlers/Processors/Batches/ClusterTransactionRequestProcessor.cs:
```diff
@@ -1,10 +1,28 @@
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Clients.Elasticsearch;
```
?

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Batches/ShardedClusterTransactionRequestProcessor.cs:
```diff
@@ -15,9 +21,16 @@ public ShardedClusterTransactionRequestProcessor(ShardedDatabaseRequestHandler r
+    {
+        token.ThrowIfCancellationRequested();
+
+        return Task.FromResult(Task.CompletedTask);
```
simply `return Task.CompletedTask`

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Batches/ShardedClusterTransactionRequestProcessor.cs:
```diff
@@ -15,9 +21,16 @@ public ShardedClusterTransactionRequestProcessor(ShardedDatabaseRequestHandler r
 
+    public override Task WaitForDatabaseCompletion(Task onDatabaseCompletionTask, long index, ClusterTransactionOptions options, ArraySegment<BatchRequestParser.CommandData> parsedCommands, CancellationToken token)
+    {
+        token.ThrowIfCancellationRequested();
```
not need?

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -3,6 +3,8 @@
 using System.Linq;
 using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
```
?

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Handlers/Processors/Batches/BatchHandlerProcessorForBulkDocs.cs:
```diff
@@ -103,8 +104,10 @@ public static async Task WaitForIndexesAsync(DocumentDatabase database, TimeSpan
                     hadStaleIndexes = true;
-
-                    await waitForIndexItem.WaitForIndexing.WaitForIndexingAsync(waitForIndexItem.IndexBatchAwaiter);
+                    using (waitForIndexItem.IndexBatchAwaiter.Register(token))
+                    {
+                        await waitForIndexItem.WaitForIndexing.WaitForIndexingAsync(waitForIndexItem.IndexBatchAwaiter);
+                    }
```
instead let's pass the token to `WaitForIndexingAsync` and do something like this in the `FrozenAwaiter` :

```
 public async Task<bool> WaitAsync(CancellationToken token)
 {
     var tcs = _tcs;
     await using (token.Register(() => tcs.TrySetCanceled()))
     {
         return await _tcs.Task.ConfigureAwait(false);
     }
 }
```

karmeli87 replied:
or maybe even better
```
return _parent.WaitAsync(token);
```

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Batches/ShardedBatchHandlerProcessorForBulkDocs.cs:
```diff
@@ -45,10 +46,11 @@ protected override async ValueTask<DynamicJsonArray> HandleTransactionAsync(Json
     {
         // no-op
         // this is passed as a parameter when we execute transaction on each shard
+        token.ThrowIfCancellationRequested();
```
no need

shaharhikri replied:
Done

ayende replied:
Still here?

shaharhikri replied:
Deleted

karmeli87 commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -891,20 +922,76 @@ private static unsafe SingleClusterDatabaseCommand ReadCommand<TTransaction>(Tra
-                    errors.Add(ToClusterTransactionErrorInfo(blittable));
+        private static DynamicJsonValue ToDynamicJsonValue(BlittableJsonReaderObject bjro)
+        {
+            var djv = new DynamicJsonValue();
+            var prop = new BlittableJsonReaderObject.PropertyDetails();
+
+            for (int i = 0; i < bjro.Count; i++)
+            {
+                bjro.GetPropertyByIndex(i, ref prop);
+                if (prop.Value is LazyStringValue lsv)
+                {
+                    djv[prop.Name] = lsv.ToString();
+                }
+                else
+                {
+                    var json = prop.Value as IDynamicJson;
+                    djv[prop.Name] = json == null ? (object)prop.Value : json.ToJson();
                 }
+            }
```
Would prefer to avoid this and work with `TryGetMember(...)`

shaharhikri replied:
Done

shaharhikri committed:
RavenDB-20628 - Addressed PR Comments

karmeli87 commented on src/Sparrow.Server/AsyncManualResetEvent.cs:
```diff
@@ -93,29 +93,37 @@ public Task<bool> WaitAsync()
             {
                 var tcs = _tcs;
-                return token.Register(() => tcs.TrySetCanceled());
+                await using (token.Register(() => tcs.TrySetCanceled()))
```
hmm when thinking about this, I'm not sure you can simply cancel the `_tcs`, others can await it... 
consider you have 2 threads doing `WaitAsync` with different token, so one will cancel the other.
you need to do `_parent.WaitAsync(token)`

shaharhikri replied:
Done, used "WithCancellation" method instead to not cancel the task, only throw and stop waiting in case the token was canceled

karmeli87 commented on src/Sparrow.Server/AsyncManualResetEvent.cs:
```diff
@@ -93,29 +93,37 @@ public Task<bool> WaitAsync()
             {
-                var waitAsync = _tcs.Task;
+                var tcs = _tcs;
+                await using (token.Register(() => tcs.TrySetCanceled()))
```
same

shaharhikri replied:
Removed

karmeli87 commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -962,22 +960,41 @@ private static DynamicJsonArray GetGeneratedResults(BlittableJsonReaderArray bjr
+                throw new InvalidOperationException("Database cluster transaction command must have a type");
+            djv[nameof(ICommandData.Type)] = type;
+
+            var lastModifiedParsed = false;
```
fix the warning

shaharhikri replied:
Fixed

shaharhikri committed:
RavenDB-20628 - Addressed PR Comments

shaharhikri committed:
RavenDB-20628 - Fix ClusterWideTransaction_WhenStore_ShouldCreateCompareExchange test

ayende commented on src/Raven.Server/Documents/Handlers/Processors/Batches/ClusterTransactionRequestProcessor.cs:
```diff
@@ -22,6 +39,68 @@ public ClusterTransactionRequestProcessor(DatabaseRequestHandler requestHandler,
+                {
+                    lastDocumentEtag = DocumentsStorage.ReadLastDocumentEtag(tx.InnerTransaction);
+                    lastTombstoneEtag = DocumentsStorage.ReadLastTombstoneEtag(tx.InnerTransaction);
+                    modifiedCollections = GetModifiedCollections(context, database, parsedCommands);
```
We need this to ensure that we wait on the right collections after failover, right?
We cannot add this to the `ClusterTransactionResult`, since that just write the actions to be executed later.

To be honest, give that this is only for failover, I would rather:
* If this is on the current node, use the information we already have
* If this is a failover node, wait for all collections

shaharhikri replied:
Done. 
I used the 'onDatabaseCompletionTask' to get the 'modifiedCollections' from the db.

shaharhikri committed:
RavenDB-20628 - Get modifiedCollections from cluster transaction task for 'wait for indexing'.
in case of failover and completed cluster transaction, we will 'wait for indexing' with all the db collections.
